File: week6/Chapter_8_Refactoring/javascript/src/afterRefactor/money.js
export class Money {
  constructor(amount, currency = "USD") {
    this.amount = amount;
    this.currency = currency;
  }
  add(other) { this.#check(other); return new Money(this.amount + other.amount, this.currency); }
  sub(other) { this.#check(other); return new Money(this.amount - other.amount, this.currency); }
  mul(factor) { return new Money(this.amount * factor, this.currency); }
  #check(other) {
    if (this.currency !== other.currency) throw new Error(`Currency mismatch: ${this.currency} vs ${other.currency}`);
  }
  fmt() { return `${this.amount.toFixed(2)} ${this.currency}`; }
}

File: week6/Chapter_8_Refactoring/python/app/after_refactor/money.py
from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True, slots=True)
class Money:
    """Tiny value object that keeps arithmetic explicit."""
    amount: float
    currency: str = "USD"

    def __add__(self, other: "Money") -> "Money":
        self._check(other)
        return Money(self.amount + other.amount, self.currency)

    def __sub__(self, other: "Money") -> "Money":
        self._check(other)
        return Money(self.amount - other.amount, self.currency)

    def __mul__(self, factor: float) -> "Money":
        return Money(self.amount * factor, self.currency)

    def _check(self, other: "Money") -> None:
        if self.currency != other.currency:
            raise ValueError(f"Currency mismatch: {self.currency} vs {other.currency}")

    def fmt(self) -> str:
        return f"{self.amount:.2f} {self.currency}"
